### ISSUE TITLE
`#:default` value is ignored and all arguments are always required

### ISSUE THREAD

**Issue by ROCKTAKEY** [arunisaac/ccwl#9]
Sorry for creating many issues in short time. I'm willing to make PR if I have enough time.

When `#:default` is provided as argument on command or workflow, the argument should be optional. However, `ccwl` requires all arguments is provided even when default value is set. 

For example,  a scheme script below: 
```scheme
(define print
  (command #:inputs (message #:type string #:default "Hello")
           #:run "echo" message
           #:outputs (printed-message #:type stdout)))

(workflow ()
          (print))
```
causes error:
```
Backtrace:
           2 (primitive-load "/home/rocktakey/sandbox/ccwl/aa.scm")
           1 (_ #<procedure 7f9adeb96300 at ice-9/eval.scm:330:13 ()> #<procedure 7f9adeb962e0 at ice-9/eval.scm:336:13 (a b)> #<procedure call-with-valu…>)
In ccwl/ccwl.scm:
   378:10  0 (_ _ _ _)

ccwl/ccwl.scm:378:10: Step `print' missing required parameters `(message)'
```

**Comment by arunisaac** [comment]
> Sorry for creating many issues in short time.

No worries at all! ccwl has kinda petered out due to no one being
interested in it. So, I'm very happy that you are interested in it and
pouring oil into its rusty machinery! :-)

I'll respond to all your issues within a few days.

> I'm willing to make PR if I have enough time.

Always welcome!

**Comment by arunisaac** [comment]
Default arguments are meant to be defaults in generated CWL files only (for example, when running CWL files using cwltool). They don't have any other special meaning within ccwl.

Default arguments in the inputs of commands only make sense when you use them in a trivial workflow like so:
```scheme
(define print
  (command #:inputs (message #:type string #:default "Hello")
                      #:run "echo" message
                      #:outputs (printed-message #:type stdout)))

print
```
For compound workflows consisting of multiple commands/steps, default arguments in inputs have no effect. You need to explicitly specify inputs as usual.
```scheme
(workflow ((message #:type string))
  (print #:message message))
```

**Comment by arunisaac** [comment]
Actually, I take back what I said. Maybe, it makes sense to have arguments with default inputs even in compound workflows with multiple steps. Let me think about it a bit.

**Comment by ROCKTAKEY** [comment]
Ok, so I will show some examples to help you think. In my opinion, I think #7 should be prioritized because it makes propagation of constant value (See last section).

# Script file
Sometimes, a script file should be included as inputs. We will use another script rarely. I would like to make the script file constant, but there is no way to do so.
```scheme
(define plot
  (command #:inputs
           (gnuplot-script #:type File
                           #:default '((class . "File")
                                       (location . "../script/example.gnuplot")))
           (input #:type File)
           (output-name #:type string)
           (title #:type string)
           #:run
           "gnuplot"
           "-e" "inputFile='$(inputs.input.path)'"
           "-e" "outputFile='$(inputs[\"output-name\"])'"
           "-e" "title='$(inputs.title)'"
           gnuplot-script
           #:outputs
           (plot-svg #:type File
                     #:binding '((glob . "$(inputs[\"output-name\"])")))))

(workflow ((input #:type File)
           (output-name #:type string)
           (title #:type string)

           ;; Also needed
           (gnuplot-script #:type File
                           #:default '((class . "File")
                                       (location . "../script/example.gnuplot"))))
          (plot #:input input
                #:output-name output-name
                #:gnuplot-script gnuplot-script))
```

# Imported workflow
For generalization, large workflow has sometimes many optional inputs which is intended to be changed rarely. Usually some arguments should not be changed in our workflow, even if the imported workflow has many optional inputs.
```scheme
;; It has 100 optional arguemnts
(define example
  (cwl-workflow "example.cwl"))

(define print
  (command #:inputs (message #:type string #:default "Hello")
                             #:run "echo" message
                             #:outputs (printed-message #:type stdout)))

(workflow (
           (num #:type int)
           ;; Need other 99 arguments even when the others are not to be changed
           )
          (pipe (example
                 #:num num
                 ;; Need other 99 arguments?
                 )
                (print #:message message)))
```

# Importing workflow generated by ccwl
It is related to #7. In ccwl, constant arguments are not allowed, so we need to put all constants to default value of workflow arguments. Because `ccwl` does not parse default values, we need to write constant values also in workflow file importing other workflow generated by ccwl. 

```scheme
;; example.scm
(define example
  (command #:inputs
           (str #:type string)
           #:run "echo" str
           #:outputs
           (out #:type stdout)))

(workflow ((str #:type string #:default "Hello"))
          (example #:str str)
          ;; This is not allowed
          ;; (example #:str "Hello")
          )
```
```scheme
(define example-workflow
  (cwl-workflow "example.cwl"))

;; ...

(workflow (
           ;; Need to be writed also here
           (str #:type string #:default "Hello"))
          (pipe 
           (example-workflow #:str str)
           ;; These are not allowed
           ;; (example-workflow)
           ;; (example-workflow #:str "hello")
           ;; ...
          ))
```

**Comment by arunisaac** [comment]
Indeed default values are essential! I have implemented it now in 900ecf10e030393653592b4200d2582b15158454. Constant values as in #7 are also important. I will implement that next.

**Comment by ROCKTAKEY** [comment]
Sorry for late reply. It works very well! thanks!
